python-api/gstswitch/server.py

```diff
@@ -22,6 +24,107 @@
+
+        # After calling start(), _proc contains a subprocess.Popen instance
+        self._proc = None
+
+        # Command to run (Array passed to subprocess.Popen)
+        self._cmd = cmd
+
+        # Logfile to write to
+        self._logtarget = logtarget
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+    def start(self):
+        """Starts up the process specified in the constructor in a thread.
+        Blocks until the process is started and has a pid assigned, but not
+        longer then 0.5s.
+        """
+
+        assert self._proc is None
+        self.log.debug("starting subprocess")
+
+        self._proc = subprocess.Popen(
+            self._cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            bufsize=0,
+            shell=False)
+
+        assert self._proc is not None
+        self.log.debug("subprocess successfully")
+
+    def terminate(self):
+        """Kills the process and waits for the thread to exit"""
+        self.log.debug("killing the subprocess")
+        if self._proc:
+            self._proc.terminate()
+
+    def poll(self):
+        """Calls poll() on the running process and returns its result or None
+        if no process is running"""
+        if self._proc:
+            return self._proc.poll()
+
+        return None
+
+    @property
+    def pid(self):
+        """Returns the pid of the running process or None
+        if no process is running"""
+        if self._proc:
+            return self._proc.pid
+
+        return None
+
+    def wait_for_output(self, match, timeout=5, count=1):
+        """Searches the output already captured from the running process for
+        match and returns immediatly if match has already been captured.
+
+        Sets up a match-request for the stdout/stderr-reader and blocks until
+        match emerges in the processes stdout/stderr but not longer then
+        timeout. If no match is found until timeout is passed, a RuntimeError
+        is raised.
+        """
+
+        self.log.debug("testing for %dx '%s' in buffer", count, match)
+        if self._buffer.count(match) >= count:
+            self.log.debug("match found, returning without reading more data")
+            return
+
+        endtime = time.time() + timeout
+        while True:
+            timeout = endtime - time.time()
+            self.log.debug("waiting for data output by subprocess (remaining time to timeout = %fs)", timeout)
+            (r, w, e) = select.select([self._proc.stdout], [], [], timeout)
+            if self._proc.stdout not in r:
+                raise RuntimeError("Timeout while waiting for match "
+                                   "'%s' %dx in the subprocess output."
+                                   % (match, count,))
+
+            self.log.debug("reading data from subprocess")
+            chunk = os.read(self._proc.stdout.fileno(), 2000).decode('utf-8')
```

Comment:
`File.read` blocks until the buffer is full. `os.read` only blocks then the buffer is empty (which is checked by the `select` before). This kind of nonblocking-read is copied from the pexpect-module.

https://github.com/pexpect/pexpect/blob/master/pexpect/pty_spawn.py#L387
https://github.com/pexpect/pexpect/blob/master/pexpect/spawnbase.py#L107